docs/contributing/MODEL_MANAGER.md

```diff
@@ -0,0 +1,1214 @@
+is recommended that you set them explicitly as shown in the above example.
+
+| **Argument**     | **Type**                     | **Default** | **Description**                           |
+|------------------|------------------------------|-------------|-------------------------------------------|
+| `config`         | InvokeAIAppConfig        | Use system-wide config  | InvokeAI app configuration object |
+| `queue`          | DownloadQueueServiceBase  | Create a new download queue for internal use  | Download queue |
+| `store`          | ModelRecordServiceBase  | Use config to select the database to open  | Config storage database |
+| `event_bus`      | EventServiceBase        | None | An event bus to send download/install progress events to |
+| `event_handlers` | List[DownloadEventHandler]  | None | Event handlers for the download queue |
+
+Note that if `store` is not provided, then the class will use
+`ModelRecordServiceBase.open(config)` to select the database to use.
+
+Once initialized, the installer will provide the following methods:
+
+#### install_job = installer.install_model()
+
+The `install_model()` method is the core of the installer. The
+following illustrates basic usage:
+
+```
+sources = [
+	Path('/opt/models/sushi.safetensors'),   # a local safetensors file
+	Path('/opt/models/sushi_diffusers/'),    # a local diffusers folder
+	'runwayml/stable-diffusion-v1-5',        # a repo_id
+    'runwayml/stable-diffusion-v1-5:vae',    # a subfolder within a repo_id
+	'https://civitai.com/api/download/models/63006', # a civitai direct download link
+	'https://civitai.com/models/8765?modelVersionId=10638',  # civitai model page
+	'https://s3.amazon.com/fjacks/sd-3.safetensors', # arbitrary URL
+]
+
+for source in sources:
+   install_job = installer.install_model(source)
+   
+source2key = installer.wait_for_installs()
+for source in sources:
+    model_key = source2key[source]
+	print(f"{source} installed as {model_key}")
+```
+
+As shown here, the `install_model()` method accepts a variety of
+sources, including local safetensors files, local diffusers folders,
+HuggingFace repo_ids with and without a subfolder designation,
+Civitai model URLs and arbitrary URLs that point to checkpoint files
+(but not to folders).
+
+Each call to `install_model()` will return a `ModelInstallJob` job, a
+subclass of `DownloadJobBase`. The install job has additional
+install-specific fields described in the next section.
+
+Each install job will run in a series of background threads using
+the object's download queue. You may block until all install jobs are
+completed (or errored) by calling the `wait_for_installs()` method as
+shown in the code example. `wait_for_installs()` will return a `dict`
+that maps the requested source to the key of the installed model. In
+the case that a model fails to download or install, its value in the
+dict will be None. The actual cause of the error will be reported in
+the corresponding job's `error` field.
```

Comment:
It returns all of the installs cumulative since the last call to `wait_for_installs()`. If there is no proceeding `wait_for_installs()` then it returns info about all the installs since the `ModelInstallService` was instantiated.

So there's no race condition, but perhaps `wait_for_installs()` should not clear the list after it is called and always return the cumulative list of installs performed over the lifetime of the application? I'm unclear on the proper use case here.

Reply:
Glad to hear that there's no race.

I'm not really sure which option is better either. Let's save that discussion for the code review for this feature when we have full context on how it's being used. This feels like it's at least close to what we want.